### PR TITLE
samples: nrf52: add die temperature demo

### PR DESCRIPTION
**Pull request by overheat**
Adding a die temperature demo specific to NORDIC nRF5 serie SoC.

Signed-off-by: Aaron Tsui <aaron.tsui@outlook.com>

Already tested on nrf52840_pca10056, nrf52_pca10040 and nrf51_pca10028

### CODE REVIEW

**Comment by nashif** [comment]
tested that on HW, works as expected.

File: samples/boards/nrf52/temp/src/main.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#include <zephyr.h>
+#include <device.h>
+#include <sensor.h>
+#include <misc/printk.h>
+
+void main(void)
+{
+	struct sensor_value temp;
+
+	struct device *dev = device_get_binding("TEMP_0");
```
Comment (bwitherspoon) [inline samples/boards/nrf52/temp/src/main.c]: CONFIG_TEMP_NRF5_NAME is defined for the binding name.

File: samples/boards/nrf52/temp/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
+project(hello_world)
```
Comment (bwitherspoon) [inline samples/boards/nrf52/temp/CMakeLists.txt]: Project name _temp_nrf52_ or similar would be more appropriate.

**Review by bwitherspoon — COMMENTED** [review]
Tested on nrf52840_pca10059. A few comments. Every sample has a README.rst. It seems to me this belongs in samples/sensor/nrf5.

File: samples/boards/nrf52/temp/sample.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+sample:
+  name: nRF5 temperature
+tests:
+  sample.sensor.hts221:
```
Comment (bwitherspoon) [inline samples/boards/nrf52/temp/sample.yaml]: This is a different sensor.

File: samples/boards/nrf52/temp/src/main.c (hunk 1 above)
Comment (overheat) [reply samples/boards/nrf52/temp/src/main.c]: Thank you for reviewing. I will change it to 'CONFIG_TEMP_NRF5_NAME'.

File: samples/boards/nrf52/temp/CMakeLists.txt (hunk 2 above)
Comment (overheat) [reply samples/boards/nrf52/temp/CMakeLists.txt]: This sample is specific to nrf5 series SoC. This temperature sensor is inside nrf5 series SoC die, and I cannot find a better place. Maybe under "samples/boards/nrf52/" is good enough for now.
What do you think?

File: samples/boards/nrf52/temp/sample.yaml (hunk 3 above)
Comment (overheat) [reply samples/boards/nrf52/temp/sample.yaml]: Thanks!  I will change it.

File: samples/boards/nrf52/temp/CMakeLists.txt (hunk 2 above)
Comment (overheat) [reply samples/boards/nrf52/temp/CMakeLists.txt]: I think it will confuse the users if move it to "sample/sensors", cause it not work for other SoC.

Comment (bwitherspoon) [reply samples/boards/nrf52/temp/CMakeLists.txt]: In this particular comment I was saying the CMake project name in `project(hello_world)` should be something like `project(temp_nrf52)`.

As for the location, what you are demonstrating / testing is the sensor from drivers/sensor/nrf5/ and not the features of a particular board. It is embedded in a family of SoCs, but I don't think people will be confused by this and I would expect to find it in samples/sensor/nrf5/. See what the assigned reviewer has to say before making any changes.

Comment (mbolivar) [reply samples/boards/nrf52/temp/CMakeLists.txt]: The project name should still be changed. I think the current location is fine.

Comment (overheat) [reply samples/boards/nrf52/temp/CMakeLists.txt]: @mbolivar thanks. For the project name, what's your suggest name?

**Comment by mbolivar** [comment]
Code is looking good; tested on nrf52840_pca10056. This sample should get a documentation file, though. Then I'd be +1 to merge.

File: samples/boards/nrf52/temp/CMakeLists.txt (hunk 2 above)
Comment (mbolivar) [reply samples/boards/nrf52/temp/CMakeLists.txt]: >  I don't think people will be confused by this and I would expect to find it in samples/sensor/nrf5/

@bwitherspoon If you feel strongly it seems easy enough to move, FWIW.

**Comment by overheat** [comment]
> Code is looking good; tested on nrf52840_pca10056. This sample should get a documentation file, though. Then I'd be +1 to merge.

Thanks. I will add a documentation file later this week.

**Comment by overheat** [comment]
@mbolivar Done. Please help to review.

**Comment by mbolivar** [comment]
> @mbolivar Done. Please help to review.

Sorry for the delay, but did you forget to push? I don't see any documentation.

**Comment by overheat** [comment]
> > @mbolivar Done. Please help to review.
> 
> Sorry for the delay, but did you forget to push? I don't see any documentation.

Oops, pushed now.

File: samples/boards/nrf52/temp/README.rst [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+ .. code-block:: console
+
+   Temperature:26.500000 C
+   Temperature:26.500000 C
+   Temperature:26.500000 C
+   Temperature:26.500000 C
+   Temperature:26.500000 C
+   Temperature:26.500000 C
+    <repeats endlessly every 1 seconds>
```
Comment (mbolivar) [inline samples/boards/nrf52/temp/README.rst]:
```suggestion
    <repeats endlessly every second>
```

File: samples/boards/nrf52/temp/README.rst [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+Overview
+********
+The Temperature Example shows how to use the temperature sensor which is
+inside die and print the measurement.
+
+
+Requirements
+************
+
+Only available on the nRF5 chipsets.
```
Comment (mbolivar) [inline samples/boards/nrf52/temp/README.rst]:
```suggestion
Only available on boards with nRF5 chips.
```

File: samples/boards/nrf52/temp/README.rst [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+
+Building and Running
+********************
+
+ This project outputs sensor data to the console. It requires the die
```
Comment (mbolivar) [inline samples/boards/nrf52/temp/README.rst]:
```suggestion
 This project outputs sensor data to the console. It measures die
```

File: samples/boards/nrf52/temp/README.rst [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ temperature data and print the measurement.
```
Comment (mbolivar) [inline samples/boards/nrf52/temp/README.rst]:
```suggestion
 temperature data and prints the measurements.
```

File: samples/boards/nrf52/temp/README.rst [hunk 8]
```diff
@@ -0,0 +1,41 @@
+.. _temp_nrf5:
+
+TEMP_nRF5: Die temperature Monitor
+########################################
+
+Overview
+********
+The Temperature Example shows how to use the temperature sensor which is
+inside die and print the measurement.
```
Comment (mbolivar) [inline samples/boards/nrf52/temp/README.rst]:
```suggestion
present in nRF5 SoC dies to print temperature data.
```

File: samples/boards/nrf52/temp/README.rst [hunk 9]
```diff
@@ -0,0 +1,41 @@
+.. _temp_nrf5:
+
+TEMP_nRF5: Die temperature Monitor
```
Comment (mbolivar) [inline samples/boards/nrf52/temp/README.rst]:
```suggestion
nRF5 Die Temperature Monitor
```

**Review by mbolivar — CHANGES_REQUESTED** [review]
Thanks! Ultimately @dbkinder will decide to approve the documentation, but I have a few suggestions from my side.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
With some minor tweaks, this sample should also work on any board with fxos8700 (e.g., frdm_k64f). I suggest moving it to `samples/sensor/die_temp`

**Comment by mbolivar** [comment]
> With some minor tweaks, this sample should also work on any board with fxos8700 (e.g., frdm_k64f). I suggest moving it to `samples/sensor/die_temp`

Why shouldn't that be done as follow-up work by users of those boards?

**Comment by MaureenHelm** [comment]
> > With some minor tweaks, this sample should also work on any board with fxos8700 (e.g., frdm_k64f). I suggest moving it to `samples/sensor/die_temp`
> 
> Why shouldn't that be done as follow-up work by users of those boards?

Of course. But let's put the sample in a more generic place now and I'll send a PR later to enable it on additional boards.

**Comment by overheat** [comment]
> With some minor tweaks, this sample should also work on any board with fxos8700 (e.g., frdm_k64f). I suggest moving it to `samples/sensor/die_temp`

Hi, this sample shows the Temperature sensor inside nRF SoC, which is like a Peripheral of the SoC, ARM core acquires data through APB bus and depends on SoC implementation. That's why I suggest keeping this sample inside samples/boards/nrf52 folder instead of samples/sensor folder.
By the contract, fxos8700 is a single component, outside of SoC(K64Fxx), and communication with SoC through I2C/SPI interface.

File: samples/boards/nrf52/temp/README.rst [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+Requirements
+************
+
+Only available on boards with nRF5 chips.
+
+
+Building and Running
+********************
+
+ This project outputs sensor data to the console. It requires die
```
Comment (dbkinder) [inline samples/boards/nrf52/temp/README.rst]: I suspect ``requires`` should be ``acquires``

File: samples/boards/nrf52/temp/sample.yaml [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+  name: nRF5 temperature
+tests:
+  sample.boards.nrf52.temp:
+    platform_whitelist: nrf52840_pca10056 nrf52_pca10040 nrf51_pca10028
+    harness: console
+    tags: sensors
+    harness_config:
+        type: one_line
+        regex:
+            - "Temperature:(.*) C"
```
Comment (dbkinder) [inline samples/boards/nrf52/temp/sample.yaml]: it's good practice to have a final newline on files

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by overheat** [comment]
@mbolivar @MaureenHelm @dbkinder 
fixed.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please rebase to remove the merge commit.

> Hi, this sample shows the Temperature sensor inside nRF SoC, which is like a Peripheral of the SoC, ARM core acquires data through APB bus and depends on SoC implementation. That's why I suggest keeping this sample inside samples/boards/nrf52 folder instead of samples/sensor folder.
By the contract, fxos8700 is a single component, outside of SoC(K64Fxx), and communication with SoC through I2C/SPI interface.

The purpose of the sensor API is to abstract away those differences from the application. In fact, I'd like to see the driver-specific accelerometer samples merged into a common sample like this.

**Pull request closed without merging**